src/lambda.ts

```diff
@@ -0,0 +1,98 @@
+import {Readable} from "stream";
+import {WriteStream} from "fs";
+import {ReadLine} from "readline";
+import {ManagedUpload} from "aws-sdk/lib/s3/managed_upload";
+import {PromiseResult} from "aws-sdk/lib/request";
+import {AWSError} from "aws-sdk/lib/error";
+import {CopyObjectOutput} from "aws-sdk/clients/s3";
+
+let fs = require('fs');
+let readline = require('readline');
+let AWS = require('aws-sdk');
+let s3 = new AWS.S3();
+
+let MaxKb = parseInt(process.env.MaxObjectSize)*1000;
+
+export async function handler(event) {
+    return Promise.all(
+        event.Records.map(record => {
+            let bucket = record.s3.bucket.name;
+            let key = record.s3.object.key;
+            let sizeKb = record.s3.object.size;
+
+            console.log(`Copying ${bucket}/${key} with size ${sizeKb}`);
+
+            if (sizeKb > MaxKb) {
+                return splitFile(bucket, key)
+            } else {
+                return copy(bucket, key)
+            }
+        })
+    )
+}
+
+/**
+ * Streams the source file, splitting it into files with size ~MaxKb and streaming each to the target bucket.
+ * Note - the result is non-deterministic because output file sizes and even the number of files may vary,
+ * but all of the data will be copied.
+ */
+function splitFile(sourceBucket: string, key: string): Promise<Array<ManagedUpload.SendData>> {
+    let buildFileName = (suffix: number) => `${key}_chunk-${suffix}`;
+
+    let fileNumber = 0;
+
+    let inputStream: Readable = s3.getObject({
+        Bucket: sourceBucket,
+        Key: key
+    }).createReadStream();
+
+    let reader: ReadLine = readline.createInterface({
+        input: inputStream
+    });
+
+    let outputStream: WriteStream = fs.createWriteStream(`/tmp/${buildFileName(fileNumber)}`);
+
+    let readerResult: Promise<Array<Promise<ManagedUpload.SendData>>> = new Promise((resolve, reject) => {
+        let s3UploadResults: Array<Promise<ManagedUpload.SendData>> = [];
+
+        let uploadToS3 = (n: number) => {
+            let readStream = fs.createReadStream(`/tmp/${buildFileName(n)}`);
+            let s3UploadResult: Promise<ManagedUpload.SendData> = s3.upload({
+                Bucket: process.env.Bucket,
+                Key: `${sourceBucket}/${buildFileName(n)}`,
+                Body: readStream
+            }).promise();
+
+            s3UploadResults.push(s3UploadResult);
+        };
+
+        reader.on('line', (line: string) => {
+            outputStream.write(line + '\n');
+
+            if (outputStream.bytesWritten > MaxKb * 1000) {
+                //Begin writing to a new file stream before closing the old one
+                let oldOutputStream = outputStream;
+                outputStream = fs.createWriteStream(`/tmp/${buildFileName(fileNumber + 1)}`);
+
+                oldOutputStream.end();
+                uploadToS3(fileNumber++);
```

Comment:
this is equivalent to 
```js
uploadToS3(fileNumber)
filterNumber += 1
```
is this what you want?

Reply:
yes, it uploads the old (completed) file to s3, then increments the counter